src/optimizer/cost_and_stats_calculator.cpp

```diff
@@ -10,127 +10,404 @@
+        if (index_searchable) {
+          output_cost_ = 0;
+        } else {
+          output_cost_ = 2;
+        }
+        return;
+      }
+
+      auto property_ = output_properties_->GetPropertyOfType(PropertyType::COLUMNS)->As<PropertyColumns>();
+      auto output_stats = generateOutputStat(table_stats, property_);
+
+      if (predicate_prop == nullptr) {
+        output_cost_ += Cost::NoConditionSeqScanCost(table_stats);
+      } else {
+        if (index_searchable) {
+          output_cost_ += updateMultipleConjuctionStats(table_stats, predicate, output_stats, true);
+        } else {
+          output_cost_ += updateMultipleConjuctionStats(table_stats, predicate, output_stats, false);
+        }
+      }
+      output_stats_ = output_stats;
+    }
+
+
+    void CostAndStatsCalculator::Visit(const PhysicalProject *) {
+      // TODO: Replace with more accurate cost
+      output_cost_ = getCostOfChildren(child_costs_);
+      PL_ASSERT(child_stats_.size() == 1);
+      auto table_stats_ptr = std::dynamic_pointer_cast<TableStats>(child_stats_.at(0));
+      if (table_stats_ptr == nullptr || table_stats_ptr->GetColumnCount() == 0) {
+        output_stats_.reset(new Stats(nullptr));
+        output_cost_ = 0;
+        return;
+      }
+      auto property_ = output_properties_->GetPropertyOfType(PropertyType::COLUMNS)->As<PropertyColumns>();
+      auto output_stats = generateOutputStat(table_stats_ptr, property_);
+
+      output_cost_ += Cost::ProjectCost(std::dynamic_pointer_cast<TableStats>(child_stats_.at(0)),
+                                        std::vector <oid_t>(), output_stats);
+      output_stats_ = output_stats;
+    }
+    void CostAndStatsCalculator::Visit(const PhysicalOrderBy *) {
+      // TODO: Replace with more accurate cost
+      PL_ASSERT(child_stats_.size() == 1);
+      output_cost_ = getCostOfChildren(child_costs_);
+      auto table_stats_ptr = std::dynamic_pointer_cast<TableStats>(child_stats_.at(0));
+      if (table_stats_ptr == nullptr || table_stats_ptr->GetColumnCount() == 0) {
+        output_cost_ = 0;
+        return;
+      }
+
+      auto property_ = output_properties_->GetPropertyOfType(PropertyType::COLUMNS)->As<PropertyColumns>();
+      auto output_stats = generateOutputStat(table_stats_ptr, property_);
+      auto sort_prop = std::dynamic_pointer_cast<PropertySort>(
+        output_properties_->GetPropertyOfType(PropertyType::SORT));
+      std::vector<std::string> columns;
+      std::vector<bool> orders;
+      for(size_t i = 0; i < sort_prop->GetSortColumnSize(); i++) {
+        auto expr = sort_prop->GetSortColumn(i);
+        if (expr->GetExpressionType() == ExpressionType::VALUE_TUPLE) {
+          auto column = std::dynamic_pointer_cast<expression::TupleValueExpression>(
+            sort_prop->GetSortColumn(i))->GetColumnName();
+
+          // In case the first column is not a tuple expression
+          if (i != 0 && columns.empty()) {
+            columns.push_back(column);
```

Comment:
Why do we push_back the same column twice if the first column is not a tuple value expression? Why do we only consider the cases that the first column is not a tuple value expression, not other non-tuple expression columns?

Reply:
Since right now the Cost::OrderByCost uses only the first column and see whether the column is a primary key and the order is ascending.  So to avoid the incorrect calculation of ```order by a+b, a ``` where a is the primary key and a+b is ignored right now, I add a fake column (order is false) first.

Reply:
I see. Could you also add some comments in the codes for this corner case?

Reply:
Sure

Reply:
Be sure to add a TODO here.